crates/pop-common/src/api.rs

```diff
@@ -0,0 +1,202 @@
+// SPDX-License-Identifier: GPL-3.0
+
+use crate::APP_USER_AGENT;
+use bytes::Bytes;
+use reqwest::IntoUrl;
+use serde::de::DeserializeOwned;
+#[cfg(test)]
+use std::collections::HashMap;
+use std::{
+	error::Error as _,
+	ops::Deref,
+	sync::{Arc, Mutex},
+	time::{SystemTime, SystemTimeError},
+};
+use thiserror::Error;
+use tokio::{
+	sync::{AcquireError, Semaphore},
+	time::{Duration, Instant},
+};
+
+/// An API client.
+pub(crate) struct ApiClient {
+	permits: Arc<Semaphore>,
+	token: Option<String>,
+	rate_limits: Arc<Mutex<RateLimits>>,
+	#[cfg(test)]
+	cache: Arc<Mutex<HashMap<String, ApiResponse>>>,
```

Comment:
Simple implementation to have request caching only when being used within tests.